### PR TITLE
[SMTChecker] Add support to constructors

### LINKED ISSUE #7143: [SMTChecker] Support constructor

### PR DESCRIPTION
**Pull request by leonardoalt**
Fixes #7143 #7144 

This includes constructor inheritance.

~~Depends on https://github.com/ethereum/solidity/pull/7440~~
Depends on https://github.com/ethereum/solidity/pull/7823

### CODE REVIEW

File: libsolidity/formal/CHC.cpp [hunk 1]
```diff
@@ -143,20 +152,14 @@ bool CHC::visit(FunctionDefinition const& _function)
 	auto functionPred = predicate(*functionEntryBlock, currentFunctionVariables());
 	auto bodyPred = predicate(*bodyBlock);
 
-	// Store the constraints related to variable initialization.
-	smt::Expression const& initAssertions = m_context.assertions();
-	m_context.pushSolver();
```
Comment (leonardoalt) [inline libsolidity/formal/CHC.cpp:168]: The new initialization allows removing this hack :)

File: libsolidity/formal/CHC.cpp [hunk 2]
```diff
@@ -591,19 +603,6 @@ vector<smt::Expression> CHC::currentBlockVariables()
 	return currentFunctionVariables() + paramExprs;
 }
 
-void CHC::clearIndices()
```
Comment (leonardoalt) [inline libsolidity/formal/CHC.cpp:636]: Moved to `SMTEncoder`

File: libsolidity/formal/CHC.h [hunk 3]
```diff
@@ -121,9 +121,6 @@ class CHC: public SMTEncoder
 	std::vector<smt::Expression> currentBlockVariables();
 
 	/// Sets the SSA indices of the variables in scope to 0.
-	/// Used when starting a new block.
-	void clearIndices();
```
Comment (leonardoalt) [inline libsolidity/formal/CHC.h:128]: Moved to `SMTEncoder`

File: libsolidity/formal/SMTEncoder.cpp [hunk 4]
```diff
@@ -39,7 +40,20 @@ bool SMTEncoder::visit(ContractDefinition const& _contract)
...
+
+	// Look for all the constructor invocations bottom up.
+	for (auto const& base: _contract.annotation().linearizedBaseContracts)
+		if (auto const& constructor =  base->constructor())
+			for (auto const& invocation: constructor->modifiers())
+			{
+				auto refDecl = invocation->name()->annotation().referencedDeclaration;
+				if (auto const& baseContract = dynamic_cast<ContractDefinition const*>(refDecl))
+				{
+					solAssert(!m_baseConstructorCalls.count(baseContract), "");
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp:59]: @chriseth This assumes that a base constructor can have at most one invocation from a derived contract. Is that true?

File: libsolidity/formal/SMTEncoder.cpp [hunk 5]
```diff
@@ -515,14 +583,16 @@ void SMTEncoder::endVisit(FunctionCall const& _funCall)
...
+	solAssert(m_currentContract, "");
 	m_context.reset();
 	m_context.pushSolver();
 	m_pathConditions.clear();
 	pushCallStack({&_function, nullptr});
 	m_uninterpretedTerms.clear();
-	resetStateVariables();
-	initializeLocalVariables(_function);
+	createStateVariables(*m_currentContract);
+	createLocalVariables(_function);
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp:657]: Now variables are only created, that is, the symbolic variables are created, but no constraints are added at this point.

**Comment by Marenz** [comment]
Needs rebasing due to conflicts

**Comment by chriseth** [comment]
Conflicts.

**Comment by leonardoalt** [comment]
Actually found a bug, still working on it.

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `2fc96b8b8d76aed9b85e67c18b8a6d139fe49831`:
```
Coding style error:
 libsolidity/formal/SMTEncoder.cpp:42: for (auto const& node: _contract.subNodes())
 libsolidity/formal/SMTEncoder.cpp:43: if (
 libsolidity/formal/SMTEncoder.cpp:47: node->accept(*this);

```
Please check that your changes are working as intended.

File: libsolidity/formal/BMC.cpp [hunk 6]
```diff
@@ -442,10 +452,6 @@ void BMC::inlineFunctionCall(FunctionCall const& _funCall)
 		// The reason why we need to pushCallStack here instead of visit(FunctionDefinition)
 		// is that there we don't have `_funCall`.
 		pushCallStack({funDef, &_funCall});
-		// If an internal function is called to initialize
-		// a state variable.
-		if (m_callStack.empty())
-			initFunction(*funDef);
```
Comment (leonardoalt) [inline libsolidity/formal/BMC.cpp:453]: Moved this above, better placement.

File: libsolidity/formal/CHC.cpp [hunk 7]
```diff
@@ -65,6 +65,15 @@ void CHC::analyze(SourceUnit const& _source)
...
 	m_variableUsage.setFunctionInlining(false);
 
+	auto boolSort = make_shared<smt::Sort>(smt::Kind::Bool);
+	auto genesisSort = make_shared<smt::FunctionSort>(
+		vector<smt::SortPointer>(),
+		boolSort
+	);
+	m_genesisPredicate = createSymbolicBlock(genesisSort, "genesis");
+	auto genesis = (*m_genesisPredicate)({});
+	addRule(genesis, genesis.name);
```
Comment (leonardoalt) [inline libsolidity/formal/CHC.cpp:75]: This block is the entry point to the entire blockchain. From here any contract can be constructed. This is needed to keep different functions from different contracts alive, without having to add each contract's constructor as a fact.

File: libsolidity/formal/SMTEncoder.cpp (hunk 4 above)
Comment (leonardoalt) [reply libsolidity/formal/SMTEncoder.cpp:59]: This is similar to how these are collected in the code generator.

**Comment by leonardoalt** [comment]
Ready for review

**Comment by ekpyron** [comment]
It needs a rebase, but I'll start a review anyways.

File: libsolidity/formal/SMTEncoder.cpp [hunk 8]
```diff
@@ -90,25 +120,25 @@ bool SMTEncoder::visit(ModifierDefinition const&)
...
-			"Assertion checker does not yet support constructors."
-		);
-	}
-	else
-	{
-		_function.parameterList().accept(*this);
-		if (_function.returnParameterList())
-			_function.returnParameterList()->accept(*this);
-		visitFunctionOrModifier();
+		auto const& contract = dynamic_cast<ContractDefinition const&>(*_function.scope());
```
Comment (ekpyron) [inline libsolidity/formal/SMTEncoder.cpp]: Either cast to ``const*`` and assert that it's not null or inline this into the next line, removing the outer brackets...

File: test/libsolidity/smtCheckerTests/complex/slither/const_state_variables.sol [hunk 9]
```diff
@@ -53,6 +53,5 @@ contract MyConc{
 // ----
 // Warning: (773-792): This declaration shadows an existing declaration.
 // Warning: (1009-1086): Function state mutability can be restricted to view
-// Warning: (874-879): Underflow (resulting value less than 0) happens here.
-// Warning: (874-879): Overflow (resulting value larger than 2**256 - 1) happens here.
+// Warning: (985-1002): Underflow (resulting value less than 0) happens here.
```
Comment (ekpyron) [inline test/libsolidity/smtCheckerTests/complex/slither/const_state_variables.sol:56]: Hm... there's not a single constructor in this... can you tell what those source locations are and why this change is valid?

File: test/libsolidity/smtCheckerTests/functions/constructor_hierarchy.sol [hunk 10]
```diff
@@ -1,6 +1,16 @@
...
+contract A is C {
+	constructor() C(2) public {
+		assert(a == 2);
+		assert(a == 3);
+	}
+}
 // ----
-// Warning: (45-72): Assertion checker does not yet support constructors.
-// Warning: (93-121): Assertion checker does not yet support constructors.
+// Warning: (166-180): Assertion violation happens here
```
Comment (ekpyron) [inline test/libsolidity/smtCheckerTests/functions/constructor_hierarchy.sol:16]: It would be really cool if we could somehow make github highlight or jump to the source locations for these - I think that every time on your SMT checks ;-). But I trust it's the correct one.

File: test/libsolidity/smtCheckerTests/functions/constructor_hierarchy_3.sol [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+		a = x;
+	}
+}
+
+contract A is B {
+	constructor(uint x) B(x) C(x + 2) public {
+		assert(a == x);
+		assert(a == x + 1);
+	}
+}
```
Comment (ekpyron) [inline test/libsolidity/smtCheckerTests/functions/constructor_hierarchy_3.sol:20]: Hm... maybe another one similar to this like:
```
pragma experimental SMTChecker;
contract C {
	uint a;
	constructor(uint x) public {
		a = x;
	}
}

contract B is C {
        uint b;
	constructor(uint x) public {
		b = a + x;
	}
}

contract A is B {
	constructor(uint x) B(x) C(x + 2) public {
		assert(a == x + 2);
		assert(b == x + x + 2);
	}
}
```
With or without letting the ``B`` constructor refer to ``a``...

Comment (ekpyron) [reply test/libsolidity/smtCheckerTests/functions/constructor_hierarchy_3.sol:20]: Generally I'm seeing more tests with constructor overriding what others did, verifying that the last one was the correct one, but not really tests for *all* constructors being executed...

File: test/libsolidity/smtCheckerTests/functions/constructor_hierarchy_diamond.sol [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+	constructor(uint x) C(x + 2) public {
+		a = x;
+	}
+}
+
+contract A is B2, B1 {
+	constructor(uint x) B2(x) B1(x) public {
+		assert(a == x);
+		assert(a == x + 1);
+	}
```
Comment (ekpyron) [inline test/libsolidity/smtCheckerTests/functions/constructor_hierarchy_diamond.sol:25]: For these diamond one's it would also be nice to set ``b1 = x + a`` and ``b2 = x + a`` (state variables of ``B1``, resp. ``B2``) instead of ``a`` in the constructors of ``B1`` and ``B2`` and then check those...

File: test/libsolidity/smtCheckerTests/inheritance/constructor_state_variable_init_diamond_middle.sol [hunk 13]
```diff
@@ -0,0 +1,22 @@
+pragma experimental SMTChecker;
+
+contract A {
+	uint x = 1;
+}
+
+contract B is A {
+	uint x = 2;
```
Comment (ekpyron) [inline test/libsolidity/smtCheckerTests/inheritance/constructor_state_variable_init_diamond_middle.sol]: I guess these will be shadowing errors in the future anyways... having ``y`` here and ``z`` in ``C``, depending on each other, would go towards my other comments about test cases...

**Review by ekpyron — CHANGES_REQUESTED** [review]
First mainly some comments about test cases... do you want to add those? If so use a separate commit for now, otherwise I won't see which ones are new :-).
And I'll need another pass through the code afterwards.

File: test/libsolidity/smtCheckerTests/functions/constructor_hierarchy.sol (hunk 10 above)
Comment (ekpyron) [reply test/libsolidity/smtCheckerTests/functions/constructor_hierarchy.sol:16]: Actually not for this PR, but generally: do you think it'd be good to repeat the ``SourceLocation.text()`` of the entire assertion?
Like ``Assertion "assert(blablabla)" failed here``?

Comment (Marenz) [reply test/libsolidity/smtCheckerTests/functions/constructor_hierarchy.sol:16]: we could write a small browser plugin to provide that highlighting functionality

File: test/libsolidity/smtCheckerTests/complex/slither/const_state_variables.sol (hunk 9 above)
Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/complex/slither/const_state_variables.sol:56]: Well, there are implicit constructors and state variable initialization, which are also part of this PR.
Yes, in `contract MyConc`, since the state variables assignments were not taken into account before, `uint should_be_constant_2 = A + 1;` used to overflow. Now that it knows that `A = 1`, it doesn't anymore.

File: test/libsolidity/smtCheckerTests/functions/constructor_hierarchy.sol (hunk 10 above)
Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/functions/constructor_hierarchy.sol:16]: @ekpyron good idea

File: test/libsolidity/smtCheckerTests/functions/constructor_hierarchy_3.sol (hunk 11 above)
Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/functions/constructor_hierarchy_3.sol:20]: Ok!

Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/functions/constructor_hierarchy_3.sol:20]: Added them

File: test/libsolidity/smtCheckerTests/functions/constructor_hierarchy_diamond.sol (hunk 12 above)
Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/functions/constructor_hierarchy_diamond.sol:25]: Added

File: test/libsolidity/smtCheckerTests/inheritance/constructor_state_variable_init_diamond_middle.sol (hunk 13 above)
Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/inheritance/constructor_state_variable_init_diamond_middle.sol]: But depending on each other how? The contracts are unrelated

**Comment by leonardoalt** [comment]
Rebased and fixed comments so far

File: libsolidity/formal/CHC.h (hunk 3 above)
Comment (ekpyron) [reply libsolidity/formal/CHC.h:128]: There's one line of comment still there that should also be removed, isn't there?

**Review by ekpyron — CHANGES_REQUESTED** [review]

**Comment by leonardoalt** [comment]
@ekpyron Yes, updated

File: libsolidity/formal/CHC.cpp [hunk 14]
```diff
@@ -521,6 +520,18 @@ void CHC::setCurrentBlock(smt::SymbolicFunctionVariable const& _block)
 	m_currentBlock = predicate(_block);
 }
 
+void CHC::setCurrentBlock(
```
Comment (leonardoalt) [inline libsolidity/formal/CHC.cpp]: The annoying this is that this function is mostly the same as the one above, just changing the last line. Not sure it's worth it to pull that part out.

**Comment by leonardoalt** [comment]
@ekpyron the last commit includes the changes we discussed today

File: libsolidity/formal/CHC.cpp [hunk 15]
```diff
@@ -94,38 +103,22 @@ bool CHC::visit(ContractDefinition const& _contract)
...
-			m_interface->declareVariable(symbVar->currentName(), symbVar->sort());
-			m_context.setZeroValue(*symbVar);
-		}
-
-		connectBlocks(constructorPred, interface());
-	}
+	m_errorPredicate = createSymbolicBlock(errorFunctionSort, "error_" + suffix);
+	m_constructorPredicate = createSymbolicBlock(constructorSort(), "implicit_constructor_" + to_string(_contract.id()));
+	auto stateExprs = currentStateVariables();
+	setCurrentBlock(*m_interfacePredicate, &stateExprs);
```
Comment (ekpyron) [inline libsolidity/formal/CHC.cpp:121]:
```suggestion
	setCurrentBlock(*m_interfacePredicate, &currentStateVariables());
```
Funnily, I think this might in fact be valid here :-)...

Comment (leonardoalt) [reply libsolidity/formal/CHC.cpp:121]: I tried that first, won't compile ;)

Comment (ekpyron) [reply libsolidity/formal/CHC.cpp:121]: I think the returned temporary should live until the end of the statement, so it should be fine, since ``setCurrentBlock`` only uses it temporarily... Let me check... not that it's really important, but still :-).

Comment (ekpyron) [reply libsolidity/formal/CHC.cpp:121]: I hate to say it, but ``boost::optional`` would have supported optional references, whereas ``std::optional`` doesn't (at least yet in C++17) :-D...

Comment (ekpyron) [reply libsolidity/formal/CHC.cpp:121]: Yeah, let's just leave it be...

**Review by ekpyron — DISMISSED** [review]

**Review by ekpyron — DISMISSED** [review]

File: Changelog.md [hunk 16]
```diff
@@ -27,6 +27,7 @@ Compiler Features:
  * SMTChecker: Add break/continue support to the CHC engine.
  * SMTChecker: Support assignments to multi-dimensional arrays and mappings.
  * SMTChecker: Support inheritance and function overriding.
+ * SMTChecker: Add support to constructors including constructor inheritance.
```
Comment (chriseth) [inline Changelog.md]: Wrong version!

**Review by chriseth — APPROVED** [review]
